### PR TITLE
[Refactor] Refactor DeepEP dispatcher

### PR DESCRIPTION
**Pull request by OrangeRedeng**
## Description
Refactor the DeepEP dispatcher to introduce structured output dtype control, replacing the
`SGLANG_DEEPEP_BF16_DISPATCH` environment variable with a `DeepEPOutputDtype` enum and
automatic detection. This is part of a broader effort to make the dispatch pipeline robust for
quantized MoE models — especially on Ascend NPU — while reducing memory overhead and
simplifying user configuration.

## Motivation

The previous dispatch pipeline had several shortcomings:

- **Hardcoded FP8 assumption.** The dispatcher unconditionally enabled FP8 dispatch for
  non-NPU, non-flashinfer-cuteDSL paths, causing incorrect results for W8A8 quant and
  BF16-pretrained models.
- **Fragile env-var workaround.** Users were forced to set
  `SGLANG_DEEPEP_BF16_DISPATCH=1` to disable FP8, but this variable was not properly
  plumbed through all code paths (e.g., `_dispatch_core` in low‑latency mode) and was
  globally scoped, leading to subtle bugs.
- **No scheme‑aware dispatch.** Quantization schemes like `ModelSlimW4A4Int4MoE` and
  NPU‑compressed‑tensors MoE methods needed manual tuning but lacked any automatic
  dtype selection.

## Modifications

A concise summary of the changes follows; see the commit history for full details.

### 1. Introduce `DeepOutputDtype` enum (`moe/utils.py`)

- **New enum** with members `BF16`, `FP8`, `INT8`, `NVFP4`.
- **Server argument** `---deepep-dispatcher-output-dtype` accepts `bf16`, `fp8`, `int8`, `nvfp4`, or
  the default `auto`.
- **Deprecated** `SGLANG_DEEPEP_BF16_DISPATCH` — removed from docs, scripts, and enforce‑
  path assert.

### 2. Automatic detection logic (`moe/utils.py`)

- **New function** `get_deepep_output_dtype()` that infers the correct output dtype from:
  1. Explicit server argument (if not `auto`).
  1. Explicit deprecated env variables.
  2. If quant_config contains input_global_scale → NVFP4
  3. If quant_config contains dispather_output_dtype → parse it
  4. If `flashinfer_cutedsl` or `is_cutlassl` runner backend → BF16 .
  5. Fallback to `BF16` on NPU.
  6. Fallback to `FP8` for standard FP8‑pretrained models.

### 3. Dispatcher refactor (`deepep.py`)

- **`_DeepEPDispatcherImplBase.__init__`** now receives `deepep_output_dtype` (as
  `DeepOutputDtype`), sets `self.use_fp8`, `self.use_nvfp4`
- **`dispatch_a` / `_dispatch_core`** in both normal and low‑latency implementations
  **no longer compute `use_fp8` locally**; they use the pre‑computed flags stored in `self`.

### 4. NPU‑specific path enhancements (`ep_moe/layer.py`)

- **Explicit `isinstance(self.scheme, (ModelSlimW4A4Int4MoE,))` check** that calls
  `torch_npu.npu_dynamic_quant(hidden_states, dst_type=torch.quint4x2)` before dispatch,
  enabling the correct BF16 output path for `ModelSlimW4A4Int4MoE` on Ascend.
- **NPU compressed‑tensors fused W4A16 MoE method** patched to pass through the proper
  `deepep_output_dtype` inferred from the scheme.
- **AWQ DeepEP** now supported on Ascend.

### 5. Additional improvements

- **`Kimi K2.5` (and some other compressed‑tensors models):** loading speed improved up to 3× on NPU.
- **CI:** updated CI to mention the new server argument
  and remove references to the deprecated env var.
- **Docs:** updated expert‑parallelism documentation to mention the new server argument
  and remove references to the deprecated env var.

### Kimi K2.5
image

### Qwen3.5-35B-A3B-w8a8-mtp

server: 
```SGLANG_SET_CPU_AFFINITY=1 PYTORCH_NPU_ALLOC_CONF=expandable_segments:True STREAMS_PER_DEVICE=32 HCCL_BUFFSIZE=1536 ASCEND_RT_VISIBLE_DEVICES=12,13,14,15 python3 -m sglang.launch_server --model-path "./weights/Qwen/Qwen3.5-35B-A3B-w8a8-mtp/" --tp 4 --mem-fraction-static 0.8 --max-total-tokens 66000 --trust-remote-code --attention-backend ascend --device npu --host 127.0.0.1 --port 30088  --moe-a2a-backend deepep --deepep-mode auto --cuda-graph-max-bs 128 --disable-radix-cache```

client:
```python ./benchmark/gsm8k/bench_sglang.py --num-questions 1319 --port 30088 --data-path ./datasets/gsm8k/test.jsonl --parallel 128```

image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

2. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
8. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): [Run #25996415014](https://github.com/sgl-project/sglang/actions/runs/25996415014)
Latest PR Test (Extra): :warning: **Not enabled** — add `run-ci-extra` label to opt in.

### CODE REVIEW

**Comment by xieminghe1** [comment]
OK, but for users, right now there’s no assertion, so it defaults to FP8.However, setting the env var export SGLANG_DEEPEP_BF16_DISPATCH=1 will disable the FP8 dispatch path.Could you add me on WeChat? My WeChat:  xieminghe0119

**Comment by OrangeRedeng** [comment]
> OK, but for users, right now there’s no assertion, so it defaults to FP8.However, setting the env var export SGLANG_DEEPEP_BF16_DISPATCH=1 will disable the FP8 dispatch path.Could you add me on WeChat? My WeChat: xieminghe0119

Due to a bug, we can't enable SGLANG_DEEPEP_BF16_DISPATCH right now, even if we specify it, we will always return a quantized FP8/int8 hidden_states from dispather. As I understand it, this issue isn't limited to Ascend

**Comment by xieminghe1** [comment]
yes, export SGLANG_DEEPEP_BF16_DISPATCH=1 hase been deleted.  you can fix the bug and restore SGLANG_DEEPEP_BF16_DISPATCH=1 functionality，but FP8_DISPATCH  may be disabled.  we may need a resonable method.

**Comment by iforgetmyname** [comment]
> yes, export SGLANG_DEEPEP_BF16_DISPATCH=1 hase been deleted. you can fix the bug and restore SGLANG_DEEPEP_BF16_DISPATCH=1 functionality，but FP8_DISPATCH may be disabled. we may need a resonable method.

I can't see a problem here. In case you need FP8_DISPATCH, just set SGLANG_DEEPEP_BF16_DISPATCH=false.

Double checked with the quant type, W4FP8 should support FP8_DISPATCH due to its activation being FP8 dtype, so the question is why in the very beginning we need this assert?

```python
        assert (
            envs.SGLANG_DEEPEP_BF16_DISPATCH.get()
        ), "W4AFP8 does not support FP8 dispatch; please set SGLANG_DEEPEP_BF16_DISPATCH=1."
```

**Comment by OrangeRedeng** [comment]
> > yes, export SGLANG_DEEPEP_BF16_DISPATCH=1 hase been deleted. you can fix the bug and restore SGLANG_DEEPEP_BF16_DISPATCH=1 functionality，but FP8_DISPATCH may be disabled. we may need a resonable method.
> 
> I can't see a problem here. In case you need FP8_DISPATCH, just set SGLANG_DEEPEP_BF16_DISPATCH=false.
> 
> Double checked with the quant type, W4FP8 should support FP8_DISPATCH due to its activation being FP8 dtype, so the question is why in the very beginning we need this assert?
> 
> ```python
>         assert (
>             envs.SGLANG_DEEPEP_BF16_DISPATCH.get()
>         ), "W4AFP8 does not support FP8 dispatch; please set SGLANG_DEEPEP_BF16_DISPATCH=1."
> ```

We described it with @xieminghe1 and we thought it would be better to create a new variable that could control the type of the dispatcher's output (nvfp4, fp8, int8, bf16, etc.).  Because it's not obvious now and is determined by input_global_scale. And everything related to SGLANG_DEEPEP_BF16_DISPATCH delete

File: python/sglang/srt/environ.py [hunk 1]
```diff
@@ -407,7 +407,6 @@ class Envs:
     SGLANG_CHUNKED_PREFIX_CACHE_THRESHOLD = EnvInt(8192)
 
     # DeepEP
-    SGLANG_DEEPEP_BF16_DISPATCH = EnvBool(False)
```
Comment (ch-wan) [inline python/sglang/srt/environ.py:414]: We'd better keep it for backward compatibility and add a deprecation message.

File: python/sglang/srt/layers/moe/utils.py [hunk 2]
```diff
@@ -148,6 +153,75 @@ def is_auto(self) -> bool:
...
+    if server_args and server_args.deepep_dispatcher_output_dtype != "auto":
+        return DeepOutputDtype(server_args.deepep_dispatcher_output_dtype)
+
+    # 1. NVFP4 is detected inside dispatch_a / _dispatch_core via quant_config; no need to infer here.
+    if self.quant_config is not None:
+        input_global_scale = self.quant_config.get("input_global_scale", None)
+        if input_global_scale is not None:
+            return DeepOutputDtype.NVFP4
+
+    if quant_scheme is not None:
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/utils.py]: We can directly pass output dtype to dispatcher's quant config. This would simplify code logic here.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 3]
```diff
@@ -57,14 +51,14 @@
     from aiter import ActivationType, QuantType
     from aiter.fused_moe import fused_moe
 elif _is_npu:
-    import torch_npu
+    pass
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: can we remove this?

Comment (OrangeRedeng) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Of course! Thanks for detecting this artifacts after the pre-commit

File: python/sglang/srt/environ.py (hunk 1 above)
Comment (OrangeRedeng) [reply python/sglang/srt/environ.py:414]: Thanks for the helpful suggestion! The variable has been restored, deprecation warnings have been added.

File: python/sglang/srt/layers/moe/utils.py (hunk 2 above)
Comment (OrangeRedeng) [reply python/sglang/srt/layers/moe/utils.py]: Thanks for the suggestion! Now the output type of the dispatcher is parsed directly from the quantization config, which reduces the amount of code

File: test/manual/layers/moe/test_moe_runners_4gpu.py [hunk 4]
```diff
@@ -64,6 +63,8 @@ class TestMoERunner4GPU(CustomTestCase):
                 "deepep",
                 "--deepep-mode",
                 "low_latency",
+                "--deepep-dispather-output-dtype",
```
Comment (ch-wan) [inline test/manual/layers/moe/test_moe_runners_4gpu.py]:
```suggestion
                "--deepep-dispatcher-output-dtype",
```

File: test/manual/test_w4a8_deepseek_v3.py [hunk 5]
```diff
@@ -192,6 +192,8 @@ def setUpClass(cls):
             "deepep",
             "--deepep-mode",
             "auto",
+            "--deepep-dispather-output-dtype",
```
Comment (ch-wan) [inline test/manual/test_w4a8_deepseek_v3.py]:
```suggestion
            "--deepep-dispatcher-output-dtype",
```

File: test/registered/4-gpu-models/test_deepseek_v3_cutedsl_4gpu.py [hunk 6]
```diff
@@ -46,12 +46,14 @@ def setUpClass(cls):
...
             "trtllm_mla",
-            "--moe-a2a-backend",
-            "deepep",
             "--moe-runner-backend",
             "flashinfer_cutedsl",
+            "--moe-a2a-backend",
+            "deepep",
             "--deepep-mode",
             "low_latency",
+            "--deepep-dispather-output-dtype",
```
Comment (ch-wan) [inline test/registered/4-gpu-models/test_deepseek_v3_cutedsl_4gpu.py]:
```suggestion
            "--deepep-dispatcher-output-dtype",
```

File: test/registered/4-gpu-models/test_deepseek_v3_cutedsl_4gpu.py [hunk 7]
```diff
@@ -117,12 +118,14 @@ def setUpClass(cls):
...
```
Comment (ch-wan) [inline test/registered/4-gpu-models/test_deepseek_v3_cutedsl_4gpu.py]:
```suggestion
            "--deepep-dispatcher-output-dtype",
```

File: python/sglang/srt/layers/moe/utils.py [hunk 8]
```diff
@@ -148,6 +154,70 @@ def is_auto(self) -> bool:
         return self == DeepEPMode.AUTO
 
 
+class DeepOutputDtype(Enum):
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/utils.py]:
```suggestion
class DeepEPOutputDtype(Enum):
```

File: docs/platforms/ascend/ascend_npu_best_practice.md [hunk 9]
```diff
@@ -248,7 +248,6 @@ do
     then
         echo "${P_IP[$i]}"
         export HCCL_BUFFSIZE=1600
-        export DEEP_NORMAL_MODE_USE_INT8_QUANT=1
```
Comment (ping1jing2) [inline docs/platforms/ascend/ascend_npu_best_practice.md:251]: i think you are supposed to do the same thing in docs_new folder

**Review by ch-wan — APPROVED** [review]
LGTM. We can merge it when it passes all CI tests.

File: python/sglang/srt/layers/moe/token_dispatcher/deepep.py [hunk 10]
```diff
@@ -328,7 +333,6 @@ def __init__(
         self.params_dtype = params_dtype
         self.deepep_mode = deepep_mode
```
[the remark this answers is not in the document]
Comment (OrangeRedeng) [reply python/sglang/srt/layers/moe/token_dispatcher/deepep.py:331]: 346 string -> self.quant_config: Optional[dict] = None

**Review by ping1jing2 — APPROVED** [review]

**Comment by ping1jing2** [comment]
I merged it as ch-wan had approved it and all GPU CIs passed
